TripalCultivate/TripalCultivate-Phenotypes
# G4.89 header validator

**Issue #89** Update header validator

## Motivation

This PR updates header validator as described in the issue.
The update will implement a validator that will check that the header row (row 1) of the data file is not an empty line/string and that values (headers) match the expected headers defined by the importer instance.



## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Create header validator


## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

ValidHeadersTest - testHeaders()
1. Test an empty headers array input.
2. Test with missing headers.
3. Test some headers in the wrong order.
4. Test a valid header.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*


```
// Create instance of the validator:
$validator_id = 'valid_headers';
$validator_instance = \Drupal::service('plugin.manager.trpcultivate_validator')
  ->createInstance($validator_id);


// Set the headers expected by the importer:
$validator_instance->setHeaders([
  [
    'name' => 'Header 0',
    'type' => 'required'  // #0
  ],
  [
    'name' => 'Header 1',
    'type' => 'required'  // #1
  ],
  [
    'name' => 'Header 2',
    'type' => 'required'  // #2
  ],
  [
    'name' => 'Header 3',
    'type' => 'optional'  // #3
  ],
  [
    'name' => 'Header 4',
    'type' => 'optional'  // #4
  ],
  [
    'name' => 'Header 5',
    'type' => 'required'  // #5
  ]
]);

// Create an headers input array:
$headers_input = [
  0 => 'Header 0',
  1 => 'Header 1',
  2 => 'Header 2',
  3 => 'Header 3',
  4 => 'Header 4',
  5 => 'Header 5',
];

// Validation will fail if you pass an empty headers array, if any required 
// headers are missing, or if the headers are not in the correct order.

// Validate headers.
$validation_status = $validator_instance->validateRow($headers_input);
dpm($validation_status);
```
